<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<title>Throwing an exception in JavaScript | </title>
		
		<link rel="stylesheet" href="../../../../../tore/@/fonts/listΣ.css?v=1">
<link rel="stylesheet" href="../../../../../tore/layd/deff/&/aefb44b7dac371a3860e9079298a950e.ructΣ.css">
<link rel="stylesheet" href="../../../../../tore/&/2290bf7aeadc4261337d6f570a3cf0c5.apprΣ.css">

<script src="../../../../../tore/&/77748427269d1c81490f5acb6ce46fc1.apprΣ.js" defer></script>


<meta name="yandex-verification" content="00eb45b7559e5c68">
<meta name="google-site-verification" content="Af0TM2OrwIEtjoQXxYz2-1C-21E1m7DB23_kWpE9sLg">
<meta name="viewport" content="width=device-width, initial-scale=1.0">		
		<meta name="description" content="In this lesson, we will learn how to throw exceptions in JavaScript.">
		<link rel="shortcut icon" type="image/x-icon" href="../../../../../favicon.ico">
	</head>
	<body id="top" data-gq="D">
		<div data-r="ruct/wrap">
			<header>
				<div data-r="ruct/topp">
	<div data-r="topp/left">
		<div data-w="nict/logo">
	<img data-e="img" data-x src="" alt="">
</div>	</div>
	<div data-r="topp/righ">
		offline version v3	</div>
	<div data-r="topp/cent">
		<br><br>	</div>
</div>								
				<div data-r="ruct/navi">
					<div data-w="nict/over">
			<span data-e="hash" data-x>⊗jsSpExcThr</span>
				<span data-e="lofo" data-x>145 of 281</span>
				<a href="../../../../../en/javascript/book/supreme.html" data-e="link" data-x>menu</a>
	</div>					<nav data-w="nict/mainmenu">
	<section data-e="main" data-x data-gm-acte="T">
		<span data-q-href="/en/markup/" data-e="open" data-x data-m-targ="markup">Markup</span>
		<span data-q-href="/en/javascript/" data-e="open" data-x data-m-targ="javascript">JavaScript</span>
	</section>
	
	<section data-e="subb" data-x data-gm-acte="F" data-m-tabb="markup">
		<a href="../../../../../en/markup/manual/html.html">HTML manual</a>
		<a href="../../../../../en/markup/manual/css.html">CSS manual</a>
	</section>
	
	<section data-e="subb" data-x data-gm-acte="F" data-m-tabb="javascript">
		<a href="../../../../../en/javascript/book/prime.html">Prime book</a>
		<a href="../../../../../en/javascript/book/practice.html">Practice book</a>
		<a href="../../../../../en/javascript/book/supreme.html">Supreme book</a>
		<a href="../../../../../en/javascript/manual.html">Manual</a>
	</section>
</nav>									</div>
			</header>
			<main>
				<div data-w="nict/hett">
	<h1>Throwing an exception in JavaScript</h1>
</div>				<div data-p="cont/pane">
<p>
	In the previous lessons, we've explored 
	two places where JavaScript throws an 
	exception when something goes wrong.
</p>
<p>
	However, there may be other situations 
	in your project that are exceptional 
	for you but not for JavaScript. In 
	this case, we can create and throw 
	our own custom exceptions.
</p>
<p>
	Let's explore the syntax required for 
	this. First, you need to create an 
	exception using the <code data-gs="comd/c" data-x>new Error</code> 
	command, passing the text of the 
	exception as a parameter:
</p>
<div data-p="bock/pane">
	<code data-w="bock/code" data-lo="high" data-m-lang="javascript">new Error('an exception text');</code>
</div>
<p>
	This exception must then be thrown 
	using the <code data-gs="comd/c" data-x>throw</code> command:
</p>
<div data-p="bock/pane">
	<code data-w="bock/code" data-lo="high" data-m-lang="javascript">throw new Error('an exception text');</code>
</div>
<p>
	Throwing an exception causes JavaScript 
	to think that an exception has occurred. 
	This means that such an exception can 
	be caught using the <code data-gs="comd/c" data-x>try-catch</code> 
	construct and handled as needed.
</p>
<p>
	Let's look how to use it at an 
	example. Suppose we have a 
	function that divides 
	one number by another:
</p>
<div data-p="bock/pane">
	<code data-w="bock/code" data-lo="high" data-m-lang="javascript">function div(a, b) {
	return a / b;
}</code>
</div>
<p>
	Let's assume that division by zero 
	is forbidden and an exception should 
	be thrown when trying to do this. To 
	do this, we will check in the function 
	if there is an attempt to divide by 
	<code data-gs="comd/n" data-x>0</code>. If not, we will divide, 
	and if yes, we will throw an exception:
</p>
<div data-p="bock/pane">
	<code data-w="bock/code" data-lo="high" data-m-lang="javascript">function div(a, b) {
	if (b !== 0) {
		return a / b;
	} else {
		throw new Error('division by zero error');
	}
}</code>
</div>
<p>
	Let's just try to divide by <code data-gs="comd/n" data-x>0</code> 
	first without catching the exception:
</p>
<div data-p="bock/pane">
	<code data-w="bock/code" data-lo="high" data-m-lang="javascript">alert( div(3, 0) ); </code>
</div>
<p>
	In this case, the script execution will 
	be interrupted and an error will appear 
	in the console with the 
	<code data-gs="comd/s" data-x>'division by zero error'</code> text 
	(check it). Let's now catch our error 
	and handle it somehow:
</p>
<div data-p="bock/pane">
	<code data-w="bock/code" data-lo="high" data-m-lang="javascript">try {
	alert( div(3, 0) );
} catch (error) {
	alert('you are trying to divide by 0, which is forbidden');
}</code>
</div>
<div data-w="cont/task">
	<p>
		In JavaScript, trying to extract the 
		square root of a negative number 
		doesn't result in an exception being thrown:
	</p>
	<div data-p="bock/pane">
		<code data-w="bock/code" data-lo="high" data-m-lang="javascript">let result = Math.sqrt(-1);
console.log(result); // shows NaN</code>
	</div>
	<p>
		Write your own function that will extract 
		the root from a number and at the same 
		time throw an exception if the root is 
		extracted from a negative number.
	</p>
</div>
</div>			</main>
				<div data-w="nict/nebr">
					<a href="../../../../../en/javascript/book/supreme/exceptions/applying-types.html" data-e="prev" data-x>
				<span data-e="arow" data-x>&larr;</span>
			</a>
							<a href="../../../../../en/javascript/book/supreme/exceptions/throwing-types.html" data-e="next" data-x>
				<span data-e="arow" data-x>&rarr;</span>
			</a>
			</div>
			<footer data-w="ruct/foot">
				<div data-w="nict/finf">
	<span data-e="text" data-x>
		Trepachev Dmitry &copy; 2012-2024	</span>
	<br>
	<a data-e="link" data-x href="https://t.me/trepachev_dmitry">t.me/trepachev_dmitry</a>
</div>							</footer>
			<div data-w="nict/jump">
	<a href="#top"><span>&uarr;</span></a>
</div>			<div data-w="nict/lagr">
	<a href="../../../../../en/javascript/book/supreme/exceptions/throwing.html" data-m-curr="T">en</a><a href="../../../../../ru/javascript/book/supreme/exceptions/throwing.html" data-m-curr="F">ru</a></div>					</div>
		
		<script src="../../../../../tore/@/metrikaΣ.js?v=1"></script>	</body>
</html>